File: src/libGEjava/Rule.java
/*
 * Rule.java
 *
 * Created on June 21, 2007, 12:06 PM
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */
package libGEjava;
/**
 *
 * @author adilraja
 */
import java.util.*;
import java.util.ArrayList;// The left-hand side of the rule

public class Rule extends ArrayList<Production>{
    private boolean recursive;// Recursive nature of rule
    private int minimumDepth;	// Minimum depth of parse tree for production to map to terminal symbol(s)
    public ArrayList<Symbol> lhs;
    /**
     *Creates a new rule with newLength elements.
     */
    public Rule(final int newLength){
        super(newLength);
	this.ensureCapacity(newLength);
	setRecursive(false);
	setMinimumDepth(32767>>1);
        lhs =new ArrayList();
}
    /**
     *Copy constructor
     */
    public Rule(final Rule copy){
        super();
        lhs=new ArrayList();
        Iterator<Symbol> lhsIt=copy.lhs.iterator();
	while(lhsIt.hasNext()){
            Symbol tmpsymb=lhsIt.next();
		lhs.add(new Symbol(tmpsymb));
	}
	setRecursive(copy.getRecursive());
	setMinimumDepth(copy.getMinimumDepth());
    }
   
/**
 *Delete all productions, and all symbols stored in lhs.
 */
 
    
public void clear(){
    try{
    lhs.clear();//clear the lhs vector
	super.clear();// Clear production vector
    }
    catch(java.lang.Exception e){
        System.out.print(e);
    }
}

    
/**
 *Return the recursive nature of this rule.
 */

public final boolean getRecursive(){
	return recursive;
}

/**
 *Update the recursive nature of this rule.
 */

public void setRecursive(final boolean newRecursive){
	recursive=newRecursive;
}

/**
 *Return the minimum mapping depth of this rule.
 */
public final int getMinimumDepth(){
	return minimumDepth;
}

/**
 * Update the minimum mapping depth of this Rule.
 */
public void setMinimumDepth(final int newMinimumDepth){
	minimumDepth=newMinimumDepth;
}


}

File: libGEC++/libGE-0.26/src/Rule.cpp
// Rule.cpp -*- C++ -*-
#ifndef _RULE_CPP_
#define _RULE_CPP_

//#include <config.h>

#include<iostream>

#include"Rule.h"

///////////////////////////////////////////////////////////////////////////////
// Creates a new rule with newLength elements.
Rule::Rule(const unsigned int newLength):vector<Production>(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Rule::Rule(unsigned int)' called\n";
#endif
	resize(newLength);
	setRecursive(false);
	setMinimumDepth(INT_MAX>>1);
}

///////////////////////////////////////////////////////////////////////////////
// Copy constructor.
Rule::Rule(const Rule &copy):vector<Production>(copy){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Rule::Rule(Rule const &copy)' called\n";
#endif
	vector<Symbol*>::const_iterator lhsIt=copy.lhs.begin();
	while(lhsIt!=copy.lhs.end()){
		lhs.push_back(new Symbol(**lhsIt));
		lhsIt++;
	}
	setRecursive(copy.getRecursive());
	setMinimumDepth(copy.getMinimumDepth());
}

///////////////////////////////////////////////////////////////////////////////
// Destructor; call redefined operator clear().
Rule::~Rule(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Rule::~Rule()' called\n";
#endif
	clear();
}

///////////////////////////////////////////////////////////////////////////////
// Delete all productions, and all symbols stored in lhs.
void Rule::clear(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'void Rule::clear()' called\n";
#endif
	vector<Symbol*>::iterator it=lhs.begin();
	while(it!=lhs.end()){
		delete *it;
		it++;
	}
	lhs.clear();// Clear lhs vector
	erase(begin(),end());// Clear production vector
}

///////////////////////////////////////////////////////////////////////////////
// Return the recursive nature of this rule.
bool Rule::getRecursive() const{
#if (DEBUG_LEVEL >= 2)
	cerr << "'bool Rule::getRecursive() const' called\n";
#endif
	return recursive;
}

///////////////////////////////////////////////////////////////////////////////
// Update the recursive nature of this rule.
void Rule::setRecursive(const bool newRecursive){
#if (DEBUG_LEVEL >= 2)
	cerr << "'void Rule::setRecursive(bool const)' called\n";
#endif
	recursive=newRecursive;
}

///////////////////////////////////////////////////////////////////////////////
// Return the minimum mapping depth of this rule.
unsigned int Rule::getMinimumDepth() const{
#if (DEBUG_LEVEL >= 2)
	cerr << "'int Rule::getMinimumDepth() const' called\n";
#endif
	return minimumDepth;
}

///////////////////////////////////////////////////////////////////////////////
// Update the minimum mapping depth of this Rule.
void Rule::setMinimumDepth(const unsigned int newMinimumDepth){
#if (DEBUG_LEVEL >= 2)
	cerr << "'int Rule::setMinimumDepth(int const)' called\n";
#endif
	minimumDepth=newMinimumDepth;
}

#endif

